Stuckmann.waw.register.renaming

F-Stuckmann opened the pull request:
This PR renames registers in loops that have a WAW dependency. Thus it aims to increase scheduling freedom for software pipelining.

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: just 2024.

F-Stuckmann replied:
changed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+//===----- AEIWarRegRewriter.cpp - Rewrite Register to not include multiple Defs
```
nit: War?

F-Stuckmann replied:
missed that in the file renaming, thanks

andcarminati commented on llvm/test/CodeGen/AIE/aie2/ra/waw_reg_renaming_loop.mir:
```diff
@@ -0,0 +1,292 @@
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+# (c) Copyright 2023-2024 Advanced Micro Devices, Inc. or its affiliates
```
nit: also just 2024.

F-Stuckmann replied:
fixed

andcarminati commented on llvm/lib/Target/AIE/AIE2RegisterInfo.cpp:
```diff
@@ -457,3 +457,33 @@ AIE2RegisterInfo::getCoveringSubRegs(const TargetRegisterClass &RC) const {
   return Subregs;
 }
+
+bool AIE2RegisterInfo::isVecOrAccRegClass(const TargetRegisterClass *RC) const {
```
Hi @F-Stuckmann, as we already encode vector types tied to register classes, I suggest a very compact implementation as follows:
```
bool AIE2RegisterInfo::isVecOrAccRegClass(const TargetRegisterClass *RC) const {
  // ******** Vector classes ********
  for (auto I = legalclasstypes_begin(*RC); I != legalclasstypes_end(*RC);
       ++I) {
    EVT VT = *I;
    if (VT.isVector())
      return true;
  }

  return false;
}
```
This can be also moved to the base class. In this case we also avoid target-specific knowledge.

F-Stuckmann replied:
accepted the suggestion and moved the method to AIEBaseRegisterinfo.h

F-Stuckmann replied:
I did not accept this change, because MVT are used earlier in the pipeline and in the register renaming pass I have access to the actual registers, which i can querry with my implementation.

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+
+namespace {
+
+cl::opt<bool> DisablePass(
```
I think it is not clear if your pass is in the optimization pipeline or not. My suggestion is to add this pass using our `AIE2PassConfig`. Also, you can move this flag to `AIE2TargetMachine.cpp` to control the insertion of the pass in the optimization pipeline. You can also use something like:

```
cl::opt<bool> EnableWAWRegRewrite(
    "aie-wawreg-rewrite",
    cl::desc("Enable the WAW Register Renaming in loops (default true)"),
    cl::init(true), cl::Hidden);
```

Using the pass manager, we can use (assuming a proper place for insertion):
```
  if (EnableWAWRegRewrite)
    addPass(createAIEWawRegRewriter());
```

Then we can disable using `-aie-wawreg-rewrite=false`.

F-Stuckmann replied:
made the pass registration clear

F-Stuckmann replied:
I moved it `AIE2TargetMachine.cpp` for better readability

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+                                   const MachineInstr *InstrB);
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+  if (DisablePass)
```
You can remove this after changing the pass creating place, or use:

```
  if (skipFunction(MF.getFunction()))
    return false;
```

To rely on function attribute (it can be independent of the optimization level) to decide about the optimization.

F-Stuckmann replied:
Since i removed the enable option to the AIE2TargetMachine, this is no longer needed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+                                     MF, VRM, LRM, MRI, LIS, TRI);
+
+          if (Replaced) {
+            MI.dump();
```
In this case, we are printing every time we enter this if. Maybe remove this dump or transform it in debug message, like `LLVM_DEBUG(dbgs() << MI);`

F-Stuckmann replied:
fixed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+    llvm::SmallVector<MCPhysReg, 16> UsedPhyRegs;
+    llvm::SmallVector<MCPhysReg, 16> UsedVirtRegs;
+    llvm::SmallVector<MCPhysReg, 16> AlreadyReplaced;
+    llvm::SmallVector<std::array<unsigned int, 2>, 8> ReplacementRegs;
```
Maybe we can use:

```
llvm::SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
```

F-Stuckmann replied:
added correct ClassTypeinfo

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+          continue;
+
+        if (isWAWCandidate(Reg, MI, UsedPhyRegs, UsedVirtRegs, VRM, MRI)) {
+          unsigned int PreSwapPhyReg = VRM.getPhys(Reg).id();
```
We can use the `MCRegister` to hide the underlying type here (instead of `unsigned int`).

F-Stuckmann replied:
added correct ClassTypeinfo

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+                         const MachineBasicBlock &MBB);
+
+static bool
+reuseSubRegs(llvm::SmallVector<const MachineInstr *, 4> &OrigInstr,
```
As we are under `llvm` namespace, we can remove `llvm::`.

F-Stuckmann replied:
removed all unnecessary llvm namespaces

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+  return Modified;
+}
+
+static llvm::SmallVector<const MachineBasicBlock *, 4>
```
I think this is a good candidate for AIELoopUtils, it can be useful for other passes.

F-Stuckmann replied:
moved it

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,466 @@
+                          << MF.getName());
+  LLVM_DEBUG(llvm::dbgs() << " ***\n");
+
+  for (const MachineBasicBlock *MBB : LoopBBs) {
```
I think this loop can be refactored as a new function, like `renameMBBPhysRegs`.

F-Stuckmann replied:
moved it

gbossu commented on llvm/test/CodeGen/AIE/aie2/ra/waw_reg_renaming_loop.mir:
```diff
@@ -0,0 +1,292 @@
+  bb.0.entry:
+  successors: %bb.1
+  liveins: $p0, $p1, $p2, $d0, $d1_3d, $d2
+  liveins: $r0
```
Nit: Can you write a single `liveins` line?

F-Stuckmann replied:
fixed

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+    "aie-waw-disable-copy-sub-reg-handling",
+    cl::desc("Disable Copy Sub register handling, where WAW regs will not be "
+             "replaced, if reg parts are reused."),
+    cl::init(false), cl::Hidden);
```
Could you add a specific test for that option? For example, we can add two different `RUN:` lines, with different `--check-prefix` FileCheck commands. I'm curious to see what case we are avoiding here. And it's generally it's great to have tests for each "special case" :)

F-Stuckmann replied:
I added a run with the prefix disabled. 
Should I denote somewhere that redundant_sub_reg_moves is the function that tests this option?

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+  // therefore should not renamed or even registered
+  assert(MI.getNumOperands() == 2 &&
+         "Copy MI does not have 2 operands, Naive Copy Elimination will "
+         "not work! Please fix.");
```
I think we can remove that assert. Verifying properties like this is the job of the `MachineVerifier`

F-Stuckmann replied:
removed

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+  if (DstReg.isVirtual())
+    DstPhyReg = VRM.getPhys(DstReg);
+  else
+    DstPhyReg = DstReg.asMCReg();
```
Is this like needed? Seems that we are converting to `MCReg`, and then back to `Register`. Same below

F-Stuckmann replied:
simplified

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+  return false;
+}
+
+bool isCopyElimination(const MachineInstr &MI, const VirtRegMap &VRM) {
```
Nit: Maybe rename to `isIdentityCopy`?

F-Stuckmann replied:
fixed

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+  return PhysRegs;
+}
+
+static bool isUndefReserved(
```
I'd be happy with a docstring :) It's not entirely clear to me what we are verifying, and for what reason

F-Stuckmann replied:
added documentation

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+          continue;
+        if (VRM.hasRequiredPhys(Op.getReg()))
+          continue;
+        if (Op.isTied())
```
I think we need to check `isRenamable()` as well. I would then assume we do not need to check `isTied()`

F-Stuckmann replied:
`isRenamable()` is only possible on physical registers. Since the pass only looks at virtual registers, this check is not possible and not relevant

gbossu replied:
Can you check how `isRenamable` is inferred then? I'd like to make sure we are not missing a special case here.

F-Stuckmann replied:
ARM has some register allocation constraints, that are not contained in operand register classes.
This function is a helper function for these cases.

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+                        << printReg(VRM.getPhys(Reg), &TRI, 0, &MRI) << " with "
+                        << printReg(PhysReg, &TRI, 0, &MRI) << '\n');
+
+      VRM.clearVirt(Reg);
```
Why not use `LRM.unassign()` to also update the interference matrix?

F-Stuckmann replied:
changed and added correct alias register blocking

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+  for (const MCPhysReg &PhysReg : RC->getRegisters()) {
+    // iterate Regs of RC and allocate first found reg that does not
+    // interfere
+    if (find(ForbiddenPhysRegs, PhysReg) != ForbiddenPhysRegs.end())
```
We cannot use a `SmallSet`?

F-Stuckmann replied:
removed, since ForbiddenPhysReg now gets updated with each Register that is replaced. Therefore during the generation of the replacement registers, the newly replaced Register will automatically be ignored.

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,455 @@
+      PhysRegs.push_back(PhysReg);
+  }
+
+  std::sort(PhysRegs.begin(), PhysRegs.end(), std::less<MCPhysReg>());
```
Why re-sort? Did you get better results compared to using the RC order?

F-Stuckmann replied:
removed, since the performance improvement was not there to justify it

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+private:
+};
+
+static bool
```
nit: with anonymous namespace you don't need static.

F-Stuckmann replied:
removed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+
+  bool runOnMachineFunction(MachineFunction &Fn) override;
+
+private:
```
nit: empty private.

F-Stuckmann replied:
removed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+    const MachineRegisterInfo &MRI, const LiveIntervals &LIS,
+    const AIEBaseRegisterInfo &TRI);
+
+static bool replaceReg(const Register Reg,
```
I think that most of those functions can be moved to the methods of the pass, and some parameters can be converted to class members. We are passing to much parameters to every function, and some of them are always the same. `const MachineRegisterInfo &MRI` for example...

F-Stuckmann replied:
I moved all the Machine function derived objects and Analysispass results to the class attributes. The methods are now less crowded with arguments.

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+//===----- AEIWaWRegRewriter.cpp - Rewrite Register to not include multiple Defs
```
I think `clang-format` broke this like because it is too long (more than 80 columns).

F-Stuckmann replied:
fixed,

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+                         const MachineRegisterInfo &MRI,
+                         const AIEBaseRegisterInfo &TRI);
+
+// Check if the UnDef is the last define for a Register, or if there is
```
nit: `UnDef ` -> `Def` .

F-Stuckmann replied:
no this is actually a undef. 
A register can also be used, when it was only written to with a undef. A unit test clears this up

andcarminati commented on llvm/lib/Target/AIE/Utils/AIELoopUtils.h:
```diff
@@ -15,11 +15,16 @@
 std::optional<int64_t> getMinTripCount(const MDNode *);
 
+SmallVector<const MachineBasicBlock *, 4>
+getLoopBBs(const MachineLoopInfo &MLI);
```
nit:  Usually we use BB for `BasicBlock` and MBB `MachineBasicBlock` it is not a convention but it is used in this way most of times.

F-Stuckmann replied:
fixed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+                       const LiveIntervals &LIS,
+                       const AIEBaseRegisterInfo &TRI);
+
+static bool isWAWCandidate(const Register &Reg, const MachineInstr &MI,
```
I suggest a more discritive name

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+          continue;
+        if (!TRI.isVecOrAccRegClass(MRI.getRegClass(Reg)))
+          continue;
+        // // Subregisters can reuse Constants in the high part of the
```
nit: double comment // //

F-Stuckmann replied:
fixed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+                           SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+                           SmallVector<MCPhysReg, 16> &UsedVirtRegs,
+                           VirtRegMap &VRM, const MachineRegisterInfo &MRI) {
+  bool FoundPhyReg = UsedPhyRegs.end() != find(UsedPhyRegs, VRM.getPhys(Reg));
```
nit: const in both cases.

F-Stuckmann replied:
fixed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+                           const AIEBaseRegisterInfo &TRI) {
+  SmallSet<MCPhysReg, 16> BlockedPhysRegs;
+  MCPhysReg PhysReg;
+  for (auto &MI : *MBB) {
```
As we are iterating with an explicit type in different parts of this code, I think we should expand this auto as well to be consistent..

F-Stuckmann replied:
fixed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+char AIEWawRegRewriter::ID = 0;
+char &llvm::AIEWawRegRewriterID = AIEWawRegRewriter::ID;
+
+INITIALIZE_PASS(AIEWawRegRewriter, "aie-waw-reg-rewrite", "AIE waw-reg rewrite",
```
nit: second parameter can be `DEBUG_TYPE`.

F-Stuckmann replied:
fixed it

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+      if (OpA.getSubReg() != OpB.getSubReg())
+        continue;
+
+      if (containEqualSubRegDefs(InstrA, InstrB))
```
It seems that this `if` is invariant to this loop.

F-Stuckmann replied:
you are right, i moved it outside

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> &UndefPartReuse,
+    const VirtRegMap &VRM) {
+  unsigned ID = Def.getReg().id();
+  if (Def.isUndef()) {
```
nit: fuse these two ifs.

F-Stuckmann replied:
fused them

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,461 @@
+static bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                         SmallVector<const MachineInstr *, 4> &OtherInstr) {
+  for (auto [InstrA, InstrB] : zip(OrigInstr, OtherInstr)) {
+    if (!InstrA->isCopyLike())
```
nit: fuse these two ifs.

F-Stuckmann replied:
fixed

gbossu commented on llvm/lib/Target/AIE/Utils/AIELoopUtils.cpp:
```diff
@@ -46,4 +46,14 @@ std::optional<int64_t> getMinTripCount(const MDNode *LoopID) {
 }
 
+SmallVector<const MachineBasicBlock *, 4>
+getLoopBBs(const MachineLoopInfo &MLI) {
```
I'd rename to `getInnermostLoopBlocks`, because you ignore outer loops

F-Stuckmann replied:
fixed it

F-Stuckmann committed:
[WAW reg] added initial unit test & pass logic

F-Stuckmann committed:
[WAW reg] copies are not renamed

F-Stuckmann committed:
[WAW reg] handle alias registers in replacement

F-Stuckmann committed:
[WAW reg] ignore non vector and accumulator registers

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,444 @@
+//===----- AEIWaWRegRewriter.cpp - Rewrite Regs to remove Defs ---------===//
```
nit: WaW -> Waw.

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,444 @@
+
+  bool Modified = renameMBBPhysRegs(LoopBBs);
+
+  this->MF = nullptr;
```
This refactor looks really nice! I think you don't need this resetting.

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,443 @@
+          MCRegister PreSwapPhyReg = VRM->getPhys(Reg).id();
+          bool Replaced = replaceReg(Reg, ForbiddenPhysRegs);
+
+          if (Replaced) {
```
nit: as we have just one use for `Raplaced`, we can use directly `if (replaceReg(...))`. Or just `const bool Replaced =...`

F-Stuckmann replied:
folded it into the if

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,443 @@
+  if (!MI.isCopy())
+    return false;
+
+  Register DstPhyReg = MI.getOperand(0).getReg();
```
nit: maybe a lambda `GetPhyReg`.

like `return GetPhyReg(DstPhyReg) == GetPhyReg(SrcPhyReg);`

See line 252.

F-Stuckmann replied:
wrote a function to extract the Physical register from the `MachineOperand`. I kept the variable names, to highlight the functionality.

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,443 @@
+                                     const MachineBasicBlock &MBB) {
+  const Register Reg = Def.getReg();
+  bool FoundDef = false;
+  for (const MachineInstr &MI : MBB) {
```
Can we simplify this code using:

`for (const MachineInstr &MI : llvm::reverse(MBB))` ?

F-Stuckmann replied:
you are write i could rewrite this function

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,443 @@
+    const Register &Reg, const MachineInstr &MI,
+    const SmallVector<MCPhysReg, 16> &UsedPhyRegs,
+    const SmallVector<MCPhysReg, 16> &UsedVirtRegs) {
+  bool FoundPhyReg = UsedPhyRegs.end() != find(UsedPhyRegs, VRM->getPhys(Reg));
```
nit: `const` in both `bool`'s

F-Stuckmann replied:
add const to identify temp variables for reader's benefit

andcarminati commented on llvm/test/CodeGen/AIE/aie2/ra/waw_reg_renaming_loop.mir:
```diff
@@ -0,0 +1,579 @@
+# RUN: llc -mtriple=aie2 -verify-machineinstrs --run-pass=greedy,aie-waw-reg-rewrite,virtregrewriter %s -o - | FileCheck %s
+# RUN: llc -mtriple=aie2 -verify-machineinstrs --run-pass=greedy,virtregrewriter %s -o - | FileCheck %s --check-prefix=DISABLE-PASS
+#
+---
```
Maybe a small description of each test can be included, forecasting the expected result.

F-Stuckmann replied:
added a comment for each test, to help clarify what each test does

andcarminati commented on llvm/test/CodeGen/AIE/aie2/ra/waw_reg_renaming_loop.mir:
```diff
@@ -0,0 +1,579 @@
+#
+# (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+
+# RUN: llc -mtriple=aie2 -verify-machineinstrs --run-pass=greedy,aie-waw-reg-rewrite,virtregrewriter %s -o - | FileCheck %s
```
I think it is better to rely on the pass pipeline itself here. For the first case:

`--start-before=greedy --stop-after=virtregrewriter`

Second case:

`--start-before=greedy --stop-after=virtregrewriter --aie-wawreg-rewrite=false`

In this way you make sure that your pass is correctly inserted within the pipeline.

F-Stuckmann replied:
fixed

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,443 @@
+  bool
+  renameMBBPhysRegs(const SmallVector<const MachineBasicBlock *, 4> &LoopBBs);
+  SmallSet<MCPhysReg, 16>
+  getBBForbiddenReplacements(const MachineBasicBlock *MBB);
```
nit: follow the same convention `getMBBForbiddenReplacements`

F-Stuckmann replied:
I also changed the `LoopMBB` for consistency's sake

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,443 @@
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  bool Modified = renameMBBPhysRegs(LoopBBs);
```
Should we collapse as `return renameMBBPhysRegs(LoopBBs);`?

F-Stuckmann replied:
yes, this was a remnant of a debug run

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,443 @@
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+
+      PhysReg = Op.getReg();
```
We can have a function for this code. A suggested a lambda for a similar case before, but we can have a function that can be called here and there.

F-Stuckmann replied:
wrote a function to extract the Physical register from the `MachineOperand`.

gbossu commented on llvm/lib/Target/AIE/AIE2RegisterInfo.cpp:
```diff
@@ -457,3 +457,32 @@ AIE2RegisterInfo::getCoveringSubRegs(const TargetRegisterClass &RC) const {
   }
   return Subregs;
 }
+bool AIE2RegisterInfo::isVecOrAccRegClass(const TargetRegisterClass *RC) const {
```
Nit: We expect the pointer to be non-null, so it is better to use a reference.

F-Stuckmann replied:
fixed

gbossu replied:
I'll re-open the discussion so I remember to check after you push the changes :)

F-Stuckmann replied:
should be fixed now. (I broke it in one of my earlier "fixups")

gbossu commented on llvm/lib/Target/AIE/AIE2RegisterInfo.cpp:
```diff
@@ -457,3 +457,32 @@ AIE2RegisterInfo::getCoveringSubRegs(const TargetRegisterClass &RC) const {
+  // ******** Vector classes ********
+  if (RC->getID() == AIE2::VEC256RegClassID ||
+      find(VEC256Superclasses, RC) != std::end(VEC256Superclasses))
+    return true;
```
If you look in the function just above, we use things like `AIE2::VEC1024RegClass.hasSubClassEq(&RC)`, which I think is a bit more readable. Could you maybe rewrite those ifs using `hasSubClassEq()`?

F-Stuckmann replied:
yeah you are right, i got confused with subclass in the sense of "C++ inherited class" and AIE subclass (X4 is a subclass/register of Y2)

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+    "aie-waw-disable-copy-sub-reg-handling",
+    cl::desc("Disable Copy Sub register handling, where WAW regs will not be "
+             "replaced, if reg parts are reused."),
+    cl::init(false), cl::Hidden);
```
Nit: I'd prefer to avoid "double negations", maybe rename to e.g. `aie-enable-waw-reg-renaming"

F-Stuckmann replied:
I will remove the option fully, since it does not make sense to ever disable it

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+  void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                    const MCPhysReg &PhysReg);
+
+  Register getPhysReg(const MachineOperand &MO) const;
```
Nit: I think `getAssignedPhysReg` would better describe what this function is doing.

F-Stuckmann replied:
fixed

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+                    << MBB->getNumber() << "\n");
+  for (auto &Reg : BlockedPhysRegs) {
+    LLVM_DEBUG(dbgs() << printReg(Reg, TRI, 0, MRI) << " ");
+  }
```
Nit: Maybe wrap that whole loop in a `LLVM_DEBUG` macro?

F-Stuckmann replied:
moved the for loop into the LLVM_DEBUG pragma

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+}
+
+SmallVector<MCPhysReg, 16> AIEWawRegRewriter::getPriorityReplacementRegs(
+    const Register Reg, const SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs) {
```
Nit: Add something like `assert(Reg.isVirtual())` to verify the precondition of this function

F-Stuckmann replied:
added the check, but the `getPriorityReplacementRegs `function is private and the only call it has, only allows Reg to be virtual. So the check should be unnecessary

gbossu replied:
I know it's unnecessary, that's why it's an assert :) The point is to verify the pre-conditions of our functions, those asserts should never trigger, but you never know how the code will evolve in the future

F-Stuckmann replied:
added an assert

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+    AU.addRequired<SlotIndexes>();
+    AU.addPreserved<SlotIndexes>();
+    AU.addRequired<LiveDebugVariables>();
+    AU.addPreserved<LiveDebugVariables>();
```
Do you really preserve it? I wonder if you need to update certain things when renaming registers.

F-Stuckmann replied:
checked and added comments to explain

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  return renameMBBPhysRegs(LoopMBBs);
```
Nit: There's a lot of indentation in `renameMBBPhysRegs`. How about doing something like
```
for (const MachineBasicBlock *MBB : LoopMBBs)
  renameMBBPhysRegs(*MBB);
```
? This way `renameMBBPhysRegs` can solely focus on renaming one BB, no need for a loop which spans the whole function.

F-Stuckmann replied:
fixed it

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+          // keep track of already visted physical and virtual Regs
+          UsedPhyRegs.push_back(VRM->getPhys(Reg));
+          UsedVirtRegs.push_back(Register::virtReg2Index(Reg));
+        }
```
There's a lot of logic in that loop, I would expect something simple like
```
if (isCandidate(...)) {
  // Then do renaming
} else {
  // Keep track of defs
}
```

Would it be possible to write the code like this? I'm wondering why some early checks have an early `continue;` statement, while those in `isWAWCandidate()` do not.

F-Stuckmann replied:
moved the renaming function call outside of the second if condition. 
this way it is clearer, that the renaming happens and only on success, i record the changes. 
Does this clear it up?

F-Stuckmann replied:
I rewrote the whole for clarity.

gbossu commented on llvm/test/CodeGen/AIE/aie2/ra/waw_reg_renaming_loop.mir:
```diff
@@ -0,0 +1,609 @@
+# RUN: llc -mtriple=aie2 -verify-machineinstrs --start-before=greedy --stop-after=virtregrewriter %s -o - | FileCheck %s
+#
+# disable pass to highlight the changes the WAW register rewriting pass performs
+# RUN: llc -mtriple=aie2 -verify-machineinstrs --start-before=greedy --stop-after=virtregrewriter --aie-wawreg-rewrite=false %s -o - | FileCheck %s --check-prefix=DISABLE-PASS
```
Nit: I would really just have the first RUN line. Having this one really just tests greedy, and makes the test cases a bit too bloated IMO. What do you think @andcarminati @martien-de-jong ?

F-Stuckmann replied:
simplified the test (and removed the disabled run). The comments should help explain what the individual unit tests are actually doing

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+        //  performance if this physical register is removed
+        if (!DisableCopySubRegHandling &&
+            isUndefReserved(Op, MI, *MBB, UndefPartReuse))
+          continue;
```
I wonder if it wouldn't just be enough to prevent renaming registers that are defined through sub-registers? Or maybe make the check more generic and preventing renaming when one lane is defined through a COPY. (Because then we might prevent coalescing)

F-Stuckmann replied:
If a sub register has a copy operation, it is immediatly ignored for renaming

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,435 @@
+  // Undef Reg is still being filled, too early to tell, if it should be
+  // removed
+  if (Def.isUndef() && !isLastDef(Def, MI, MBB))
+    return true;
```
Note: `undef` really only appears on the first partial def. If a register is defined through e.g. 4 sub-regs, then only the first assignment will get undef, not the "middle ones". I guess what you want to check here is rather `if (Def.getSubReg() && !isLastDef(...))`.

But as said in https://github.com/Xilinx/llvm-aie/pull/168/files#r1743706575, I wonder if we could not just prevent renaming altogether when a vreg is defined through a subreg.

F-Stuckmann replied:
I rewrote the whole for clarity.

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,430 @@
+  bool Modified = false;
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  SmallVector<MCPhysReg, 16> UsedPhyRegs;
```
nit: I feel that `UsedPhyRegs` and  `UsedVirtRegs`  are good candidates for a bitset/bitvector.

F-Stuckmann replied:
fixed with a Bitvector

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,371 @@
+  const bool AlreadyUsedPhysReg = UsedPhyRegs[VRM->getPhys(Reg)];
+  const bool SubRegCopy = hasSubRegCopy(Reg.id(), SubRegDefs);
+
+  return MRI->getRegClass(Reg) && AlreadyUsedPhysReg && !MI.isCopy() &&
```
Just a question: when `MRI->getRegClass(Reg)` can return null?

andcarminati replied:
nit: maybe you can test the trivial things first and then finalize with:

```
return !hasSubRegCopy(Reg.id(), SubRegDefs);
```

F-Stuckmann replied:
The check was unnecessary, since i only work on Virtual registers. 
I also simplified the structure

gbossu replied:
I think it's a good point, we seem to check that `Reg.isVirtual()` before entering this function. How about making this an official pre-condition with e.g. `assert(Reg.isVirtual() && MRI->getRegClass(Reg))`?

F-Stuckmann replied:
I also added another assert to the `replaceReg` function

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,371 @@
+        continue;
+      
+      MCRegister PreSwapPhyReg = VRM->getPhys(Reg).id();
+      if (isWAWCandidate(Reg, MI, UsedPhyRegs, SubRegDefs) &&
```
Nit: I feel this check is more about profitability, compared to the checks above that are about "legality". Maybe we can rename that function to `isWorthRenaming()`?

F-Stuckmann replied:
changed the name

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,371 @@
+
+  bool renameMBBPhysRegs(const MachineBasicBlock *MBBs);
+  SmallSet<MCPhysReg, 16>
+  getMBBForbiddenReplacements(const MachineBasicBlock *MBB);
```
Could you add a docstring for getMBBForbiddenReplacements? I will forget what is considered "forbidden" in a couple of weeks :)

F-Stuckmann replied:
added a comment to clarify its intention

F-Stuckmann commented:
I just resolvd the merge conflict and modified tests, so that each individual commit will not cause llvm-lit test failures.

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -94,8 +94,7 @@ class AIEWawRegRewriter : public MachineFunctionPass {
   // pass removes.
-  SmallSet<MCPhysReg, 16>
-  getMBBForbiddenReplacements(const MachineBasicBlock *MBB);
+  SmallSet<MCPhysReg, 16> getDefBlockedPhysRegs(const MachineBasicBlock *MBB);
```
Super-nit: "defBlocked" is a bit unclear. Maybe just `getDefinedPhysRegs`?

F-Stuckmann replied:
updated the name

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,378 @@
+  BitVector UsedPhyRegs(TRI->getNumRegs());
+  SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
+
+  SmallSet<MCPhysReg, 16> ForbiddenPhysRegs = getDefinedPhysRegs(MBB);
```
nit: this sounds like another candidate to BitVector.

F-Stuckmann replied:
changed to Bitvector

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,378 @@
+  std::map<unsigned, SmallVector<const MachineInstr *, 4>> SubRegDefs;
+  for (const MachineInstr &MI : MBB) {
+    for (const MachineOperand Def : MI.defs()) {
+      unsigned ID = Def.getReg().id();
```
nit: the unsigned operator is is the same as id(), do you can remove .id(), the result is the same. I think you can also use `MCPhysReg` instead on `unsigned`.

F-Stuckmann replied:
I should have made it clear, that here I just want virtual registers. Since this map is used to check, if any part of a Virtual Register was defined by a `copy `instruction. Changed the datatype and also added a check in the population of the map to make this distinction clear.

F-Stuckmann commented:
1. Simplified the blocking check, if a Subregister contains a copy instruction
3. squashed all fixups

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,270 @@
+                                  const BitVector &BlockedPhysRegs);
+
+  bool isWorthRenaming(const Register &Reg, const MachineInstr &MI,
+                       const BitVector &UsedPhyRegs) const;
```
nit: UsedPhysReg?

F-Stuckmann replied:
changed every occurence of PhyReg to PhysReg

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,270 @@
+
+  for (const MachineInstr &MI : *MBB) {
+    for (const MachineOperand &Op : MI.defs()) {
+      addAliasRegs(BlockedPhysRegs, getAssignedPhysReg(Op));
```
getAssignedPhysReg can return NoRegister. Is that ok for addAliasReg?

F-Stuckmann replied:
already handeled in a fixup

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,270 @@
+            std::make_pair(PreSwapPhyReg, VRM->getPhys(Reg)));
+
+      } else {
+        // keep track of already visted physical Regs
```
nit: *visited*

F-Stuckmann replied:
fixed it typo

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,347 @@
+          replaceReg(Reg, BlockedPhysRegs)) {
+
+        LLVM_DEBUG(dbgs() << MI);
+        Modified = true;
```
This is modifying the local variable, not the class member that is representing the changes the pass made.

F-Stuckmann replied:
I removed the local variable from the loop

martien-de-jong commented on llvm/lib/Target/AIE/Utils/AIELoopUtils.cpp:
```diff
@@ -86,6 +86,16 @@ getDedicatedFallThroughPreheader(const MachineBasicBlock &LoopBlock) {
+SmallVector<const MachineBasicBlock *, 4>
+getInnermostLoopBlocks(const MachineLoopInfo &MLI) {
+  SmallVector<const MachineBasicBlock *, 4> LoopBBs;
+  for (const MachineLoop *ML : MLI) {
```
I expect this `for` only visits the outer loops

F-Stuckmann replied:
fixed, I reused `isSingleMBBLoop` and created a unit test for the SingleMBBLoop block detection

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,347 @@
+        continue;
+
+      MCPhysReg PreSwapPhysReg = VRM->getPhys(Reg);
+      if (isWorthRenaming(Reg, UsedPhysRegs, VRegWithCopies) &&
```
Since UsedPhysregs is basic-block local, the effect it has on multi-block loops is questionable. Perhaps make it absolutely clear that it is intended for single block loops by only running it on single block loops.

F-Stuckmann replied:
I modified `AIELoopUtils::getSingleBlockLoopMBBs` so that it only returns a single MBB which conforms with `singleMBBLoops`

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,347 @@
+    // phi nodes that will be eliminated, do not define a physical register
+    // since they will be eliminated in a later pass, and the copy will be
+    // removed
+    if (isIdentityCopy(MI))
```
The comment speaks about phi nodes, I don't see the connection.

F-Stuckmann replied:
addressed it with a comment

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,347 @@
+  SmallVector<std::pair<MCPhysReg, MCPhysReg>, 8> ReplacementRegs;
+
+  BitVector BlockedPhysRegs = getDefinedPhysRegs(MBB);
+  SmallVector<Register, 16> VRegWithCopies = getVRegWithCopies(*MBB);
```
I don't see why this needs to be a global set. The VReg is SSA, so can only be defined by one MI. I expect we can just widen the isIdentityCopy condition.

martien-de-jong replied:
Ehh right. We are not strictly SSA after greedy.

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,347 @@
+  // dump Reassignment
+  LLVM_DEBUG(dbgs() << "WAW Replacement Overview for bb." << MBB->getNumber();
+             dbgs() << "\n"; for (std::pair<MCPhysReg, MCPhysReg> &RegPair
+                                  : ReplacementRegs) {
```
tip: `for (auto &[Reg, Replacement] : ReplacementRegs)`

F-Stuckmann replied:
fixed

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,347 @@
+  return DstPhysReg == SrcPhysReg;
+}
+
+bool AIEWawRegRewriter::isLastDef(const MachineOperand &Def,
```
This is the last def in this block. There may be other defs in other blocks, which may or may not be the actual last def. so I don't see the significance. Some comment explaining why we need to test this would be nice.

F-Stuckmann replied:
added a comment and example to clarify why it is needed

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,347 @@
+
+      } else {
+        // keep track of already visited physical registers
+        UsedPhysRegs[VRM->getPhys(Reg)] = true;
```
I think that the replacement register should also be registered as used?

F-Stuckmann replied:
addressed in a comment

andcarminati commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,347 @@
+//===----- AEIWawRegRewriter.cpp - Rewrite Regs to remove Defs ---------===//
```
nit: AIE

F-Stuckmann replied:
fixed

andcarminati commented on llvm/lib/Target/AIE/Utils/AIELoopUtils.cpp:
```diff
@@ -86,6 +86,16 @@ getDedicatedFallThroughPreheader(const MachineBasicBlock &LoopBlock) {
+getInnermostLoopBlocks(const MachineLoopInfo &MLI) {
+  SmallVector<const MachineBasicBlock *, 4> LoopBBs;
+  for (const MachineLoop *ML : MLI) {
+    if (ML->isInnermost())
```
How about considering only hardware loops here?

F-Stuckmann replied:
I synchronized with Martien, what the `MBBs `are that we are looking for. The `MBBs` I am looking for are already checked with `isSingleMBBLoop `, so my function just returns a list of the found MBBs within a `MachineFunction `

F-Stuckmann commented:
- Only Single MBB Loops are considered, in line with software pipeliner
- added a dedicated Single MBB Loop detection unit test
- I added Comments everywhere

gbossu commented on llvm/lib/Target/AIE/AIE2TargetMachine.cpp:
```diff
@@ -51,6 +51,11 @@ static cl::opt<bool>
+
+static cl::opt<bool> EnableWAWRegRewrite(
+    "aie-wawreg-rewrite",
+    cl::desc("Enable the WAW Register Renaming in loops (default true)"),
```
Nit: drop ` (default true)`, I think it will be printed in the help message because you define a default

F-Stuckmann replied:
It will not be printed in the help message, but no other help message contains the default, so i will harmonize it.

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+
+///
+/// This rewrites Registers, if they are redefined in the same Loop body and
+/// could cause WAW issues in the loop pipelining.
```
Super nit: I would maybe rephrase to make the message a bit more general, and less prone to getting outdated. E.g. `This rewrites physical register assignments in critical parts of the code (like loops) to break WAW and WAR dependencies.`

F-Stuckmann replied:
added

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+  LiveIntervals *LIS = nullptr;
+  bool Modified = false;
+
+  bool renameMBBPhysRegs(const MachineBasicBlock *MBBs);
```
Could some of those functions be `const`?

gbossu replied:
I believe they cannot be because they update `Modified`? I'd be in favor of dropping that attribute if it means we can keep all helper methods const. (Apart from `runOnFunction` of course)

F-Stuckmann replied:
`replaceReg `modifies `VirtRegMap `and `LiveRegMatrix`, thus `renameMBBPhysRegs` cannot be const. I added const to the functions, that are const.

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+  // Get all the defined physical registers that the MachineBasicBlock already
+  // uses. These physical registers should not be used for replacement
+  // candidates, since this would introduce new WAW dependencies, which this
+  // pass tries to remove.
```
Please use Doxygen's `///` instead of `//`

F-Stuckmann replied:
fixed

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  Register Reg = MO.getReg();
+
+  if (Reg.isStack())
```
Curious: What does this mean that a register is `stack`?

F-Stuckmann replied:
Registers can be `virtual`, `physical `and `stack slots`. 

The register class here is misused to also store `frame indices`, i.e. call stack locations for spilling registers during register allocation.

gbossu replied:
I'm generally wary of things I haven't seen in MIR. Do you have maybe a small test that somehow relies on the `Reg.isStack()` case?

I understand when a Register is assigned to a stack slot in `VRM`, but it's still confusing to me that a register **is** a stack slot itself.

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+  BitVector UsedPhysRegs(TRI->getNumRegs());
+
+  // record the register replacements for debugging purposes.
+  SmallVector<std::pair<MCPhysReg, MCPhysReg>, 8> ReplacementRegs;
```
Nit: Could we dump info about the renaming immediately? I'd rather avoid variables that are only useful for debugging

F-Stuckmann replied:
I already dump the renaming information, it was just a nice way of debugging large functions, where the unsuccessful replacements were crowding out the successful ones.

I removed it.

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+        continue;
+      if (MO.isTied())
+        continue;
+      if (!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg))))
```
Could you fold this `!TRI->isVecOrAccRegClass(*(MRI->getRegClass(Reg)))` check inside `isWorthRenaming()`? This way we clearly separate the legality checks from the "worthiness" checks.

F-Stuckmann replied:
moved

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+
+      } else {
+        // keep track of already visited physical registers
+        UsedPhysRegs[VRM->getPhys(Reg)] = true;
```
Do we also need to extend `UsedPhysRegs` for the `if` case?

F-Stuckmann replied:
no, because the `UsedPhysRegs` are used to identify when a physical register is a renaming candidate, by checking when it is being written to twice.

1. I do not reuse replaced registers, so they cannot cause WAW issues with later replacements.
2. During the pass setup, I already blocked all the used physical register of the Basic Block. Therefore I cannot introduce a WAW with my replacement. 

To clarify I added a comment to this section to explain it.

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+    const SmallVector<Register, 16> &VRegWithCopies) const {
+  assert(Reg.isVirtual() && MRI->getRegClass(Reg));
+
+  if (!UsedPhysRegs[VRM->getPhys(Reg)])
```
Nit: add a comment like `// Only rename registers mapped to a phys reg assigned more than once`

F-Stuckmann replied:
added clarification comment

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+}
+
+void AIEWawRegRewriter::addAliasRegs(BitVector &BlockedPhysRegs,
+                                     const MCPhysReg &PhysReg) {
```
Nit: I think `MCPhysReg` is just a 16-bit number. Pass it by value?

F-Stuckmann replied:
yes the reference would be larger, removed

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,381 @@
+  // % 0.sub_256_hi:vec512 = VLDA_dmw_lda_w_ag_idx_imm _,_                    \
+  // %7:mxm, %8:el = VMAX_LT_D8 %0, $x6                                       \
+  //
+  bool isLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
```
Super nit: I'd maybe call that `isLastVRegDef()`, because unlike some other function, here we do not look at phys reg assignments.

F-Stuckmann replied:
changed.

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+  /// $x4 = Copy $x4                                                           \
+  /// This Copy instruction will be removed in a later pass, if the source and
+  /// destination are equal, additionally it is not the first define of $x4 and
+  /// would identify a false WAW hazard, that will be removed by a later pass.
```
I think this comment repeats itself. Also, the name perfectly describes its implementation. The rationale of applying this check belongs at the place where it is called. Finally, the definition is in the same file as this declaration and fits in a few lines. I wouldn't mind just defining it here.

F-Stuckmann replied:
removed.

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+  bool isIdentityCopy(const MachineInstr &MI) const;
+
+  /// MachineInstructions can write into sub-registers (%0:sub_256_lo and
+  /// %0:sub_256_hi). In the VirtualRegisterMap, they they block the same
```
nit: *they they*

F-Stuckmann replied:
fixed

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+
+  /// MachineInstructions can write into sub-registers (%0:sub_256_lo and
+  /// %0:sub_256_hi). In the VirtualRegisterMap, they they block the same
+  /// register (i.e. $x0). Which can cause issues, when detecting Write
```
nit: 'This can cause issues when ...'

F-Stuckmann replied:
fixed

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+
+MCPhysReg
+AIEWawRegRewriter::getAssignedPhysReg(const MachineOperand &MO) const {
+  Register Reg = MO.getReg();
```
nit: This assumes that it is a register operand. Add it as a precondition to the declaration?

qcolombet replied:
Alternatively, we could just pass a Register here.

F-Stuckmann replied:
The function now takes a Register as input to make the relationship clear

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+  LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+             dbgs() << "\n");
+  bool Modified = false;
+  // Add the used physical registers one machine instruction at at time.
```
nit: *at a*

F-Stuckmann replied:
fixed

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+      } else {
+        // Keep track of already visited physical registers.
+        // Incrementally add the encountered physical registers, so that a
+        // second occurance of a physical register can trigger the register
```
nit: *occurrence*

F-Stuckmann replied:
fixed

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+        // second occurance of a physical register can trigger the register
+        // rewriting
+        // Blocked registers for the replacement are recorded in
+        // BlockedPhysRegs. I initially block all the used physical registers
```
nit: 'We' preferred over 'I'.  Using passive sentences avoid personal pronouns whatsoever.

F-Stuckmann replied:
changed to passive (european style)

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+      continue;
+
+    LiveRegMatrix::InterferenceKind IK =
+        LRM->checkInterference(LIS->getInterval(Reg), PhysReg);
```
capture LIS->getInterval(Reg)  outside of this loop?

F-Stuckmann replied:
moved it outside

martien-de-jong commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,379 @@
+                                      const MachineInstr &DefInstr,
+                                      const MachineBasicBlock &MBB) const {
+  const Register Reg = Def.getReg();
+  for (const MachineInstr &MI : llvm::reverse(MBB)) {
```
The use of this function makes the algorithm quadratic in the size of the block

F-Stuckmann replied:
precalculated a map between Register and last Defining `MachineInstr `, so the complexity does not explode

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+  LiveRegMatrix *LRM = nullptr;
+  LiveIntervals *LIS = nullptr;
+
+  bool renameMBBPhysRegs(const MachineBasicBlock *MBBs);
```
Nit: MBBs => MBB.

F-Stuckmann replied:
renamed

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+  /// definition of a virtual register to count the definitions, so that
+  /// writing into subregisters does not already trigger the register
+  /// replacement. In this case $x0 would have already been replaced, even
+  /// though there is no real WAW dependency.
```
I don't think it was said before and if it was I missed it, but that may be worth calling out here or somewhere else that the reason you are doing this is because you detect that there is a WAW when there are more than one def and sub-regs can create more than one (partial) def which are by construction not WAW.

What I'm saying is your comment makes sense when the reader has this context in their head and without it, it may be hard to understand why we do that.

qcolombet replied:
Looking through more of the PR, I don't actually see that you don't do anything special with the subregs.
I see that we only do the rewriting when we reach the last definition of a virtual reg that has been defined at least twice, but both definitions can be partial (i.e., subreg) ones.

Am I missing something?
Put differently, if I'm right, this comment is confusing since it doesn't reflect the reality.

qcolombet replied:
Talked with @F-Stuckmann offline, I missed the use of virtual registers and physical registers in the checks that happens after the call to getLastVRegDef.

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+  /// %7:mxm, %8:el = VMAX_LT_D8 %0, $x6                                       \
+  ///                                                                          \
+  /// returns a mapping between a virtual register and its last defining machine
+  /// instruction.
```
After my comment (https://github.com/Xilinx/llvm-aie/pull/168/files#r1774838323) I don't know why it is useful to have the last definition yet. I.e., why the counting and the last definition are relevant/connected here.
I'll find out more as I read the code, but wanted to flag this for now.

F-Stuckmann replied:
resolved this offline

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+  ///                                                                          \
+  /// returns a mapping between a virtual register and its last defining machine
+  /// instruction.
+  std::map<Register, const MachineInstr *>
```
FYI, you could use a densely packed map (`IndexedMap` for instance) by leveraging the Register::virtReg2Index helper method.

F-Stuckmann replied:
updated the data structure

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+  /// Block every sub- and super-register of a physical register, so that it is
+  /// removed for future replacement strategies, i.e. block wl4, wh4, y2 if X4
+  /// is used.
+  void addAliasRegs(BitVector &BlockedPhysRegs, const MCPhysReg PhysReg) const;
```
Haven't looked at the implementation yet, but this is usually more efficient to track the register units (`MCRegUnit`) instead of the `MCPhysReg` directly.

F-Stuckmann replied:
will create a draft branch for this implementation

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+        continue;
+
+      if (MI.isCopyLike() && find(VRegWithCopies, Reg) == VRegWithCopies.end())
+        VRegWithCopies.push_back(Reg);
```
You could use a SetVector instead of doing the checks manually that `Reg` is not already inserted.
You could even use a BitVector if you use the virtreg2index "trick". (Then index2virtreg when you use this bitvector)
(I'm guessing the SetVector may be more memory efficient in this case, but that's a gut feeling)

F-Stuckmann replied:
went with a `BitVector` since ordering is irrelevant (which disqualifies SetVector)

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+bool AIEWawRegRewriter::isWorthRenaming(
+    const Register &Reg, const BitVector &UsedPhysRegs,
+    const SmallVector<Register, 16> &VRegWithCopies) const {
+  assert(Reg.isVirtual() && MRI->getRegClass(Reg));
```
By construction, virtual registers must have a register class at this stage or the Machine IR is broken.

F-Stuckmann replied:
removed that assert

F-Stuckmann replied:
removed that assertion

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+}
+
+bool AIEWawRegRewriter::isIdentityCopy(const MachineInstr &MI) const {
+  if (!MI.isCopy())
```
Since AIE overloaded this method, you may want to use `TargetInstrInfo::isCopyInstr` here instead of checking for a generic copy instruction.

F-Stuckmann replied:
changed to `TargetInstrInfo::isCopyInstr`

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+      if (!Reg.isVirtual())
+        continue;
+
+      if (MI.isCopyLike() && find(VRegWithCopies, Reg) == VRegWithCopies.end())
```
Same comment as https://github.com/Xilinx/llvm-aie/pull/168/files#r1774901324, you may want to use `TargetInstrInfo::isCopyInstr`.

F-Stuckmann replied:
changed to `TargetInstrInfo::isCopyInstr`

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+    return false;
+
+  LLVM_DEBUG(dbgs() << "     WAW Replacement: Virtual Register "
+                    << Register::virtReg2Index(Reg) << " will replace "
```
Nit: you could use printReg for consistency here.

F-Stuckmann replied:
changed it

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+  /// pass tries to remove.
+  BitVector getDefinedPhysRegs(const MachineBasicBlock *MBB) const;
+
+  bool replaceReg(const Register Reg, BitVector &BlockedPhysRegs);
```
NIt: comment on what the returned `bool` means

F-Stuckmann replied:
added a comment to clarify

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+        continue;
+      // wait until the virtual register is fully defined to check if it is a
+      // replacement candidate
+      if (LastVRegDef[Reg] != &MI)
```
I think the comment could be that we don't care about several definitions of the same virtual register because even if we were to rename the virtual register, by construction we would rename all the definition and achieving nothing wrt WAW :).
I.e., it's not just about fully defined.

F-Stuckmann replied:
updated the comment

qcolombet commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,386 @@
+  return VRegWithCopies;
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(const MachineBasicBlock *MBB) {
```
The heuristic I see here is "if a register is used in a copy don't change it".
Spoke with Fabian offline about it.
The current heuristic would still rename a register used in a copy if there are several inner loops in the kernel.

At this point, we assume this is not something we care about.

One quick fix could be to keep track of the copies for all the single loop BBs instead of computing the list of copies on the fly for each MBB.

F-Stuckmann replied:
added a fixup comment

qcolombet reviewed (commented):
Few comments here and there but looks good to me!

F-Stuckmann committed:
[WAW reg] sub register definitions are not renamed

F-Stuckmann commented:
incorporated Quentins comments

qcolombet reviewed (commented):
Looks even more good to me :).
(Not sure it is appropriate for me to approve, so I leave it at that :))

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,414 @@
+
+  // Initialize the IndexedMap size
+  if (MaxVRegIndex > 0)
+    LastVRegDef.grow(MaxVRegIndex);
```
Nit: I think we could resize that map once and for all using `MRI.getNumVirtRegs()`

F-Stuckmann replied:
fixed it

gbossu commented on llvm/test/CodeGen/AIE/aie2/ra/waw_reg_renaming_loop.mir:
```diff
@@ -0,0 +1,635 @@
+# if parts of the sub-register (%281.sub_256_hi:vec512, %282.sub_256_hi:vec512)
+# are constant. Thus the "constant" subregister ($wh2) can be shared across both
+# super registers ($x2 -> %281 and %282). Otherwise an unnecessary move would be
+# introduced to move the constant in the newly renamed register.
```
Nit: I think the comment is slightly out of date as we have a more generic solution now

F-Stuckmann replied:
updated the comment

gbossu commented on llvm/test/CodeGen/AIE/aie2/ra/waw_reg_renaming_loop.mir:
```diff
@@ -0,0 +1,635 @@
+...
+
+# Don't rename superregisters on each definition of the subregisters of the same
+#  register, i.e. register %281 is defined twice within subregisters
```
Super-nit: i'd rather say `defined twice through its sub-registers`

F-Stuckmann replied:
updated the language

gbossu commented on llvm/test/CodeGen/AIE/aie2/ra/waw_reg_renaming_loop.mir:
```diff
@@ -0,0 +1,635 @@
+...
+
+# Ignore vector registers (%130:er), that are not dedicated vector/accumulator
+# registers.
```
What does it test extra compared to the test above?

F-Stuckmann replied:
a gpr register can have a superclass of a boolean vector register

In a previous implementation of vector identification, I would misidentify gpr registers. This test was designed to detect those cases.

gbossu replied:
I don't understand `superclass of a boolean vector register`. At that point of the flow, there are no "types" anymore, we just look at classes. What do you mean by a gpr register that is a "dedicated vector/accumulator register"?

F-Stuckmann replied:
In a previous version: 
if i query all the alias Reg classes, and check if any one of them is a vector, I get a `isVector=true` for this case, which is why it is a dedicated unit test. 

Of course with our current vector check implementation, this case does not appear anymore.

gbossu replied:
I think I would remove that test then, as there is no notion of "1 RC maps to 1 type". Rather, a class can accommodate multiple types. But that's important in ISel, here this late in the flow, bringing back types adds more confusion IMO.

F-Stuckmann replied:
I removed it for clarity

gbossu commented on llvm/lib/Target/AIE/AIEWawRegRewriter.cpp:
```diff
@@ -0,0 +1,408 @@
+  MCPhysReg ReplacementPhysReg = getReplacementPhysReg(Reg, BlockedPhysRegs);
+
+  if (!MCRegister::isPhysicalRegister(ReplacementPhysReg))
+    return false;
```
Super-nit: I think it means we could not find a replacement register. So we could make the check more explicit by either:
 - Having `getReplacementPhysReg` return a `std::optional`
 - Or checking `ReplacementPhysReg == MCRegister::NoRegister` explicitely

F-Stuckmann replied:
went with the `ReplacementPhysReg == MCRegister::NoRegister` check

gbossu reviewed (dismissed):
LGTM and great work! That really wasn't an easy piece of code to write 💪

F-Stuckmann commented:
fixed the conflicts
